File: agenkit-java/src/main/java/io/agenkit/skills/AgentSkill.java
package io.agenkit.skills;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;

/**
 * Represents a single agent skill loaded from a directory.
 *
 * <p>A skill directory must contain a {@code SKILL.md} file structured as:
 * <pre>{@code
 * ---
 * name: skill-name
 * description: What this skill does.
 * license: Apache-2.0  # optional
 * metadata:            # optional
 *   key: value
 * ---
 * # Skill Title
 * Markdown instructions here.
 * }</pre>
 */
public final class AgentSkill {

    private final String name;
    private final String description;
    private final String instructions;
    private final String license;
    private final Map<String, Object> metadata;
    private final Path skillDir;

    public AgentSkill(
            String name,
            String description,
            String instructions,
            String license,
            Map<String, Object> metadata,
            Path skillDir) {
        this.name = name;
        this.description = description;
        this.instructions = instructions;
        this.license = license;
        this.metadata = metadata != null
                ? Collections.unmodifiableMap(new LinkedHashMap<>(metadata))
                : Collections.emptyMap();
        this.skillDir = skillDir;
    }

    /**
     * Load a skill from a directory containing a {@code SKILL.md} file.
     *
     * @param skillDir path to the skill directory
     * @return the loaded skill
     * @throws IllegalArgumentException if the directory lacks SKILL.md, has
     *         invalid frontmatter, or is missing required fields (name, description)
     */
    public static AgentSkill fromDirectory(Path skillDir) {
        Path skillFile = skillDir.resolve("SKILL.md");
        if (!Files.exists(skillFile)) {
            throw new IllegalArgumentException("No SKILL.md found in " + skillDir);
        }

        String raw;
        try {
            raw = Files.readString(skillFile, StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new UncheckedIOException("failed to read " + skillFile, e);
        }

        // Split on "---" delimiters into at most 3 parts. File must start with "---".
        String[] parts = raw.split("---", 3);
        if (parts.length < 3) {
            throw new IllegalArgumentException(
                    "Invalid SKILL.md in " + skillDir + ": missing frontmatter delimiters");
        }

        String frontmatterText = parts[1].strip();
        String instructions = parts[2].strip();

        Map<String, Object> fm = parseFrontmatter(frontmatterText);

        Object name = fm.get("name");
        if (name == null || name.toString().isEmpty()) {
            throw new IllegalArgumentException(
                    "Missing required field 'name' in " + skillDir + "/SKILL.md");
        }

        Object description = fm.get("description");
        if (description == null || description.toString().isEmpty()) {
            throw new IllegalArgumentException(
                    "Missing required field 'description' in " + skillDir + "/SKILL.md");
        }

        Object license = fm.get("license");
        Object metadata = fm.get("metadata");

        @SuppressWarnings("unchecked")
        Map<String, Object> metadataMap = metadata instanceof Map
                ? (Map<String, Object>) metadata
                : Collections.emptyMap();

        return new AgentSkill(
                name.toString(),
                description.toString(),
                instructions,
                license != null ? license.toString() : null,
                metadataMap,
                skillDir);
    }

    /**
     * Minimal YAML frontmatter parser. Supports flat {@code key: value} pairs
     * and a single level of nested mapping under a key with an empty value
     * (e.g. {@code metadata:}) whose entries are indented. Quotes around scalar
     * values are stripped.
     */
    private static Map<String, Object> parseFrontmatter(String text) {
        Map<String, Object> result = new LinkedHashMap<>();
        Map<String, Object> nested = null;
        String[] lines = text.split("\n", -1);

        for (String line : lines) {
            if (line.isBlank() || line.strip().startsWith("#")) {
                continue;
            }

            boolean indented = line.startsWith("  ") || line.startsWith("\t");
            String trimmed = line.strip();
            int colon = trimmed.indexOf(':');
            if (colon < 0) {
                continue;
            }

            String key = trimmed.substring(0, colon).strip();
            String value = trimmed.substring(colon + 1).strip();

            if (indented && nested != null) {
                nested.put(key, parseScalar(value));
                continue;
            }

            if (value.isEmpty()) {
                // Begin a nested mapping (e.g. "metadata:").
                nested = new LinkedHashMap<>();
                result.put(key, nested);
            } else {
                nested = null;
                result.put(key, parseScalar(value));
            }
        }

        return result;
    }

    private static Object parseScalar(String value) {
        if (value.length() >= 2
                && ((value.startsWith("\"") && value.endsWith("\""))
                || (value.startsWith("'") && value.endsWith("'")))) {
            return value.substring(1, value.length() - 1);
        }
        return value;
    }

    /**
     * Render the skill as a prompt block for injection into agent messages.
     *
     * @return formatted string with skill name, description, and instructions
     */
    public String toPrompt() {
        return "# Skill: " + name + "\n\n"
                + "## Description\n" + description + "\n\n"
                + "## Instructions\n" + instructions + "\n";
    }

    public String getName() { return name; }
    public String getDescription() { return description; }
    public String getInstructions() { return instructions; }
    public Optional<String> getLicense() { return Optional.ofNullable(license); }
    public Map<String, Object> getMetadata() { return metadata; }
    public Optional<Path> getSkillDir() { return Optional.ofNullable(skillDir); }

    @Override
    public String toString() {
        return "AgentSkill{name='" + name + "', description='" + description + "'}";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AgentSkill s)) return false;
        return name.equals(s.name)
                && description.equals(s.description)
                && instructions.equals(s.instructions)
                && java.util.Objects.equals(license, s.license)
                && metadata.equals(s.metadata)
                && java.util.Objects.equals(skillDir, s.skillDir);
    }

    @Override
    public int hashCode() {
        return java.util.Objects.hash(name, description, instructions, license, metadata, skillDir);
    }
}

File: agenkit-scala/src/main/scala/io/agenkit/skills/AgentSkill.scala
package io.agenkit.skills

import java.nio.file.{Files, Path}
import scala.io.Source
import scala.util.Using

/** Represents a single agent skill loaded from a directory.
  *
  * A skill directory must contain a `SKILL.md` file structured as:
  * {{{
  * ---
  * name: skill-name
  * description: What this skill does.
  * license: Apache-2.0  # optional
  * metadata:            # optional
  *   key: value
  * ---
  * # Skill Title
  * Markdown instructions here.
  * }}}
  */
case class AgentSkill(
  name: String,
  description: String,
  instructions: String,
  license: Option[String] = None,
  metadata: Map[String, Any] = Map.empty,
  skillDir: Option[Path] = None
):
  /** Render the skill as a prompt block for injection into agent messages. */
  def toPrompt: String =
    s"# Skill: $name\n\n" +
      s"## Description\n$description\n\n" +
      s"## Instructions\n$instructions\n"

object AgentSkill:
  /** Load a skill from a directory containing a `SKILL.md` file.
    *
    * @throws IllegalArgumentException
    *   if the directory lacks `SKILL.md`, has invalid frontmatter, or is
    *   missing required fields (name, description).
    */
  def fromDirectory(skillDir: Path): AgentSkill =
    val skillFile = skillDir.resolve("SKILL.md")
    if !Files.exists(skillFile) then
      throw new IllegalArgumentException(s"No SKILL.md found in $skillDir")

    val raw = Using.resource(Source.fromFile(skillFile.toFile, "UTF-8"))(_.mkString)

    // Split on "---" delimiters. File must start with "---".
    val parts = splitOnDelimiter(raw, "---", limit = 3)
    if parts.length < 3 then
      throw new IllegalArgumentException(
        s"Invalid SKILL.md in $skillDir: missing frontmatter delimiters"
      )

    val frontmatterText = parts(1).strip()
    val instructions    = parts(2).strip()

    val fm = parseFrontmatter(frontmatterText)

    val name = fm.get("name").collect { case s: String => s }.filter(_.nonEmpty)
    if name.isEmpty then
      throw new IllegalArgumentException(s"Missing required field 'name' in $skillDir/SKILL.md")

    val description = fm.get("description").collect { case s: String => s }.filter(_.nonEmpty)
    if description.isEmpty then
      throw new IllegalArgumentException(
        s"Missing required field 'description' in $skillDir/SKILL.md"
      )

    val license = fm.get("license").collect { case s: String => s }
    val metadata = fm.get("metadata") match
      case Some(m: Map[?, ?]) => m.asInstanceOf[Map[String, Any]]
      case _                  => Map.empty[String, Any]

    AgentSkill(
      name = name.get,
      description = description.get,
      instructions = instructions,
      license = license,
      metadata = metadata,
      skillDir = Some(skillDir)
    )

  /** Split `text` on `delimiter`, producing at most `limit` parts (mirrors
    * Python's `str.split(sep, maxsplit)` where `limit == maxsplit + 1`).
    */
  private def splitOnDelimiter(text: String, delimiter: String, limit: Int): Array[String] =
    val parts  = scala.collection.mutable.ArrayBuffer.empty[String]
    var start  = 0
    var splits = 0
    var idx    = text.indexOf(delimiter, start)
    while idx >= 0 && splits < limit - 1 do
      parts += text.substring(start, idx)
      start = idx + delimiter.length
      splits += 1
      idx = text.indexOf(delimiter, start)
    parts += text.substring(start)
    parts.toArray

  /** Minimal YAML frontmatter parser supporting the subset used by skills:
    * top-level `key: value` scalars plus a single nested mapping (e.g.
    * `metadata:` followed by indented `key: value` lines). Quotes are stripped.
    */
  private def parseFrontmatter(text: String): Map[String, Any] =
    val result = scala.collection.mutable.LinkedHashMap.empty[String, Any]
    val arr    = text.split("\n", -1)
    var i      = 0
    while i < arr.length do
      val line    = arr(i)
      val trimmed = line.strip()
      if trimmed.isEmpty || trimmed.startsWith("#") then i += 1
      else if isTopLevel(line) then
        val (key, value) = splitKeyValue(trimmed)
        if value.isEmpty then
          // Possible nested mapping: gather following indented lines.
          val nested = scala.collection.mutable.LinkedHashMap.empty[String, Any]
          var j      = i + 1
          while j < arr.length && (arr(j).strip().isEmpty || isIndented(arr(j))) do
            val nl = arr(j).strip()
            if nl.nonEmpty && !nl.startsWith("#") then
              val (nk, nv) = splitKeyValue(nl)
              nested(nk) = stripQuotes(nv)
            j += 1
          if nested.nonEmpty then result(key) = nested.toMap
          else result(key) = stripQuotes(value)
          i = j
        else
          result(key) = stripQuotes(value)
          i += 1
      else i += 1
    result.toMap

  private def isTopLevel(line: String): Boolean =
    line.nonEmpty && !line.charAt(0).isWhitespace

  private def isIndented(line: String): Boolean =
    line.nonEmpty && line.charAt(0).isWhitespace

  private def splitKeyValue(line: String): (String, String) =
    val idx = line.indexOf(':')
    if idx < 0 then (line.strip(), "")
    else (line.substring(0, idx).strip(), line.substring(idx + 1).strip())

  private def stripQuotes(value: String): String =
    if value.length >= 2 &&
      ((value.startsWith("\"") && value.endsWith("\"")) ||
        (value.startsWith("'") && value.endsWith("'")))
    then value.substring(1, value.length - 1)
    else value
